src/array_value.cpp

```diff
@@ -620,4 +620,61 @@ ValuePtr ArrayValue::reverse_in_place(Env *env) {
     return this;
 }
 
+ValuePtr ArrayValue::concat(Env *env, size_t argc, ValuePtr* args) {
+    assert_not_frozen(env);
+
+    for (size_t i = 0; i < argc; i++) {
+        auto arg = args[i];
+        if (!arg->is_array()) {
+            env->raise("TypeError", "no implicit conversion of {} into Array", arg->klass()->class_name_or_blank());
+            return nullptr;
+        }
+
+        concat(*arg->as_array());
+    }
+
+    return this;
+}
+
+ValuePtr ArrayValue::rindex(Env *env, ValuePtr object, Block *block) {
+    // TODO make dry since this code is almost identical to index
+    assert(size() <= NAT_INT_MAX);
+    auto length = static_cast<nat_int_t>(size());
+    if (block) {
+        for (nat_int_t i = length - 1; i >= 0; i--) {
+            auto item = m_vector[i];
+            ValuePtr args[] = { item };
+            auto result = NAT_RUN_BLOCK_AND_POSSIBLY_BREAK(env, block, 1, args, nullptr);
+            if (result->is_truthy())
+                return ValuePtr::integer(i);
+        }
+        return NilValue::the();
+    } else if (object) {
+        for (nat_int_t i = length - 1; i >= 0; i--) {
+            auto item = m_vector[i];
+            ValuePtr args[] = { object };
+            if (item.send(env, "==", 1, args)->is_truthy())
+                return ValuePtr::integer(i);
+        }
+        return NilValue::the();
+    } else {
+        // TODO
+        env->ensure_block_given(block);
+        NAT_UNREACHABLE();
+    }
+}
+
+ValuePtr ArrayValue::none(Env *env, size_t argc, ValuePtr *args, Block *block) {
+    auto Enumerable = GlobalEnv::the()->Object()->const_fetch(SymbolValue::intern("Enumerable"))->as_module();
+    auto any_method = Enumerable->find_method(env, SymbolValue::intern("none?"));
+    return any_method->call(env, this, argc, args, block);
+}
+
+ValuePtr ArrayValue::one(Env *env, size_t argc, ValuePtr *args, Block *block) {
+    auto Enumerable = GlobalEnv::the()->Object()->const_fetch(SymbolValue::intern("Enumerable"))->as_module();
+    auto any_method = Enumerable->find_method(env, SymbolValue::intern("one?"));
+    return any_method->call(env, this, argc, args, block);
+}
```

Comment:
Calling through to the version of `one?` and `none?` on Enumerable is totally fine for now, but we should add a comment here saying it's not exactly correct. Because that's not what Ruby does.

```ruby
class Thing
  include Enumerable

  def each
    yield 1
    yield 2
    yield 3
  end
end

t = Thing.new
a = [1, 2, 3]

p t.method(:one?).owner # => Enumerable
p a.method(:one?).owner # => Array


# works
p a.one? { |i| i == 2 } # => true

# works
p t.one? { |i| i == 2 } # => true

module Enumerable
  def one?
    false
  end
end

# works
p a.one? { |i| i == 2 } # => true

# broken
p t.one? { |i| i == 2 } # => false
```

Reply:
Again, don't worry about fixing this -- what you have here is still a huge improvement, so let's ship it!

Reply:
Added the TODOs
(Still had a couple more changes that's why it was still a draft)
So let's ship it :boat: